### PR TITLE
feat/callback-lr_schedule

### PR DESCRIPTION
**Pull request by cxzhang4**

### CODE REVIEW

File: R/CallbackSetLRScheduler.R [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    param_set = ps(
+      T_max = p_int(tags = c("train", "required")),
+      eta_min = p_dbl(default = 0, lower = 0, tags = "train"),
+      last_epoch = p_int(default = -1, tags = "train"),
+      verbose = p_lgl(default = FALSE, tags = "train")
+    ),
+    id = "lr_scheduler",
+    label = "Learning Rate Scheduler",
+    man = "mlr3torch::mlr_callback_set.lr_scheduler",
+    additional_args = torch::lr_cosine_annealing
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:81]:
```suggestion
    additional_args = list(.scheduler = torch::lr_cosine_annealing)
```

File: R/CallbackSetLRScheduler.R [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      lr_lambda = p_uty(tags = c("train"), custom_check = function(x) check_class_or_list(x, "function")), # TODO: assert fn or list of fns
+      last_epoch = p_int(default = -1, lower = -1, tags = "train"),
+      verbose = p_lgl(default = FALSE, tags = "train")
+    ),
+    id = "lr_scheduler",
+    label = "Learning Rate Scheduler",
+    man = "mlr3torch::mlr_callback_set.lr_scheduler",
+    additional_args = lr_lambda
+  )
+})
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:99]: For the custom callback, we infer the parameters as the formalArgs` of the scheduling function.


```r
as_lr_scheduler = function(x) {
  # x is a lr scheduler function
  TorchCallback$new(
    callback_generator = CallbackSerLRScheduler,
    param_set = inferps(x),
    ...,
    additional_args = list(.scheduler = x)
    )
    ```
    
    If we had a scheduler that has arguments `a, b`, `inferps` creates a parameter set with two parameters `a` and `b` that are both of type `ParamUty`.

File: R/CallbackSetLRScheduler.R [hunk 3]
```diff
@@ -0,0 +1,203 @@
+#' @title Learning Rate Scheduling Callback
+#'
+#' @name mlr_callback_set.lr_scheduler
+#'
+#' @description
+#' Changes the learning rate based on the schedule specified by a `torch::LRScheduler`.
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:6]: [`torch::lr_scheduler`]

File: R/CallbackSetLRScheduler.R [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    #' @description
+    #' Creates the scheduler using the optimizer from the context
+    on_begin = function() {
+      # TODO: check that the .scheduler_args do not have the cb prefix (pretty sure this is true)
+      self$scheduler = invoke(self$scheduler_fn, optimizer = self$ctx$optimizer, .args = private$.scheduler_args)
+    },
+    #' @description
+    #' Depending on the scheduler, step after each epoch
+    on_epoch_end = function() {
+      # TODO: ensure that this happens after optimizer$step()
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:38]: this is true

File: R/CallbackSetLRScheduler.R [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+      last_epoch = p_int(default = -1, tags = "train")
+    ),
+    id = "lr_scheduler",
+    label = "Learning Rate Scheduler",
+    man = "mlr3torch::mlr_callback_set.lr_scheduler",
+    additional_args = list(.scheduler = torch::lr_step)
+  )
+})
+
+as_lr_scheduler = function(x) {
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:185]: add `per_epoch` argument

File: R/CallbackSetLRScheduler.R [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    #' The `torch` function that creates a learning rate scheduler
+    scheduler_fn = NULL,
+    #' @field scheduler (`LRScheduler`)\cr
+    #' The learning rate scheduler wrapped by this callback
+    scheduler = NULL,
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function(.scheduler, ...) {
+      self$scheduler_fn = .scheduler
+      private$.scheduler_args = list(...)
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:27]: self$on_epoch_end = function() ...

File: R/CallbackSetLRScheduler.R (hunk 3 above)
Comment (cxzhang4) [reply R/CallbackSetLRScheduler.R:6]: And list all of the supported schedulers with hyperlinks

File: R/CallbackSetLRScheduler.R [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+mlr3torch_callbacks$add("lr_scheduler_cosine_annealing", function() {
+  TorchCallback$new(
+    callback_generator = CallbackSetLRScheduler,
+    param_set = ps(
+      T_max = p_int(tags = c("train", "required")),
+      eta_min = p_dbl(default = 0, lower = 0, tags = "train"),
+      last_epoch = p_int(default = -1, tags = "train"),
+      verbose = p_lgl(default = FALSE, tags = "train")
+    ),
+    id = "lr_scheduler",
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:82]: id should match the key (i.e. unique)

File: R/CallbackSetLRScheduler.R [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    label = "Learning Rate Scheduler",
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:83]: different label for each scheduler

File: R/CallbackSetLRScheduler.R [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    if (some(!map_lgl(x, test_class(x, classname)))) {
+      return(paste0("One of the elements is not of class ", classname))
+    }
+  } else {
+    return(paste0("x must be of class ", classname, " or a list of objects of class ", classname))
+  }
+}
+
+#' @include TorchCallback.R
+mlr3torch_callbacks$add("lr_scheduler_cosine_annealing", function() {
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:73]: lr_scheduler_... -> lr_...

File: R/CallbackSetLRScheduler.R [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+# TODO: refactor to operate on batches
+#' @include TorchCallback.R
+mlr3torch_callbacks$add("lr_scheduler_one_cycle", function() {
+  TorchCallback$new(
+    callback_generator = CallbackSetLRScheduler,
+    param_set = ps(
+      max_lr = p_dbl(tags = "train"),
+      total_steps = p_int(default = NULL, tags = "train"),
+      epochs = p_int(default = NULL, tags = "train"),
+      steps_per_epoch = NULL,
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:130]: TODO: fix this

File: R/CallbackSetLRScheduler.R [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+as_lr_scheduler = function(x) {
+  assert_class(x, "lr_scheduler_generator")
+
+  TorchCallback$new(
+    callback_generator = CallbackSetLRScheduler,
+    param_set = inferps(x),
+    id = "lr_scheduler",
+    label = "Learning Rate Scheduler",
+    man = "mlr3torch::mlr_callback_set.lr_scheduler",
```
Comment (cxzhang4) [inline R/CallbackSetLRScheduler.R:193]: id, label

File: R/CallbackSetLRScheduler.R [hunk 12]
```diff
@@ -184,10 +184,12 @@ as_lr_scheduler = function(x, step_on_epoch) {
   assert_class(x, "lr_scheduler_generator")
   assert_flag(step_on_epoch)
 
+  class_name = setdiff(class(x), c("lr_scheduler", "lr_scheduler_generator"))
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:187]:
```suggestion
  class_name = class(x)[1L]
```

File: tests/testthat/test_CallbackSetLRScheduler.R [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            epochs = n_epochs, batch_size = 150, neurons = 10,
+            measures_train = msrs(c("classif.acc", "classif.ce"))
+  )
+  gamma = 0.5
+  step_size = 2
+  mlp$param_set$set_values(cb.lr_step.gamma = gamma)
+
+  mlp$param_set$set_values(cb.lr_step.step_size = step_size)
+  # doesn't pass bc step_size required param is set in the learner's paramset
+  expect_torch_callback(cb, check_paramset = FALSE)
```
Comment (cxzhang4) [inline tests/testthat/test_CallbackSetLRScheduler.R:23]: The `expect_torch_callback()` function calls `cb$generate()`, which checks that required parameters have been set. Setting the parameter like in line 21 in the learner's paramset does not work here. Is that an issue?

However, the following work
```
cb = t_clbk("lr_step", step_size = 2)
```

```
cb = t_clbk("lr_step")
cb$param_set$set_values(step_size = 2)
```

Comment (sebffischer) [reply tests/testthat/test_CallbackSetLRScheduler.R:23]: you just need to specify the required parameters to some values

Comment (cxzhang4) [reply tests/testthat/test_CallbackSetLRScheduler.R:23]: Meaning set the parameter value before I pass the callback to the learner?

Comment (sebffischer) [reply tests/testthat/test_CallbackSetLRScheduler.R]: the callback is cloned when passing it to the learner, do changing the parameter via the mlp does not change the original callback

Comment (cxzhang4) [reply tests/testthat/test_CallbackSetLRScheduler.R]: Thanks yeah it wasn't clear to me yesterday but makes sense now!

File: R/CallbackSetLRScheduler.R [hunk 14]
```diff
@@ -3,16 +3,22 @@
...
+#' Changes the learning rate based on the schedule specified by a `torch::lr_scheduler`.
+#'
+#' As of this writing, the following are available: [torch::lr_cosine_annealing()], [torch::lr_lambda()], [torch::lr_multiplicative()], [torch::lr_one_cycle()],
+#' [torch::lr_reduce_on_plateau()], [torch::lr_step()], and custom schedulers defined with [torch::lr_scheduler()].
+#'
+#' If the `lr_scheduler` steps on batches, `last_epoch` should be used to indicate the index of the last batch.
 #'
 #' @param .scheduler (`function`)\cr
-#'   The torch scheduler constructor function (e.g. `torch::lr_step`).
+#'   The `torch`` scheduler generator (e.g. `torch::lr_step`).
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:14]:
```suggestion
#'   The `torch` scheduler generator (e.g. `torch::lr_step`).
```

File: R/CallbackSetLRScheduler.R [hunk 15]
```diff
@@ -3,16 +3,22 @@
 #' @name mlr_callback_set.lr_scheduler
 #'
 #' @description
-#' Changes the learning rate based on the schedule specified by a `torch::LRScheduler`.
+#' Changes the learning rate based on the schedule specified by a `torch::lr_scheduler`.
+#'
+#' As of this writing, the following are available: [torch::lr_cosine_annealing()], [torch::lr_lambda()], [torch::lr_multiplicative()], [torch::lr_one_cycle()],
+#' [torch::lr_reduce_on_plateau()], [torch::lr_step()], and custom schedulers defined with [torch::lr_scheduler()].
+#'
+#' If the `lr_scheduler` steps on batches, `last_epoch` should be used to indicate the index of the last batch.
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:11]:
```suggestion
```

Comment (sebffischer) [reply R/CallbackSetLRScheduler.R:11]: or improve the wording

File: R/CallbackSetLRScheduler.R [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+#'
+#' @description
+#' Changes the learning rate based on the schedule specified by a `torch::lr_scheduler`.
+#'
+#' As of this writing, the following are available: [torch::lr_cosine_annealing()], [torch::lr_lambda()], [torch::lr_multiplicative()], [torch::lr_one_cycle()],
+#' [torch::lr_reduce_on_plateau()], [torch::lr_step()], and custom schedulers defined with [torch::lr_scheduler()].
+#'
+#' If the `lr_scheduler` steps on batches, `last_epoch` should be used to indicate the index of the last batch.
+#'
+#' @param .scheduler (`function`)\cr
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:13]:
```suggestion
#' @param .scheduler (`lr_scheduler_generator`)\cr
```

File: R/CallbackSetLRScheduler.R [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+#'   The `torch`` scheduler generator (e.g. `torch::lr_step`).
+#' @param ... (`list`)\cr
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:15]:
```suggestion
#' @param ... (any)\cr
```

File: R/CallbackSetLRScheduler.R [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+    scheduler_fn = NULL,
+    #' @field scheduler (`LRScheduler`)\cr
+    #' The learning rate scheduler wrapped by this callback
+    scheduler = NULL,
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function(.scheduler, step_on_epoch, ...) {
+      assert_class(.scheduler, "lr_scheduler_generator")
+      self$scheduler_fn = .scheduler
+      private$.scheduler_args = list(...)
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:34]:
```suggestion
      private$.scheduler_args = list(...)
      assert_flag(step_on_epoch)
```

File: R/CallbackSetLRScheduler.R [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+  )
+)
+
+# some of the schedulers accept lists
+# so they can treat different parameter groups differently
+check_class_or_list = function(x, classname) {
+  if (test_class(x, classname)) {
+    return(TRUE)
+  } else if (is.list(x)) {
+    if (some(!map_lgl(x, test_class(x, classname)))) {
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:58]: `check_list(x, types = classname)`

File: R/CallbackSetLRScheduler.R [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+  ),
+  private = list(
+    .scheduler_args = NULL
+  )
+)
+
+# some of the schedulers accept lists
+# so they can treat different parameter groups differently
+check_class_or_list = function(x, classname) {
+  if (test_class(x, classname)) {
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:55]:
```r
if (is.list(x)) check_list(x, types = classname) else check_class(x, classname)
```

File: R/CallbackSetLRScheduler.R [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+  TorchCallback$new(
+    callback_generator = CallbackSetLRScheduler,
+    param_set = ps(
+      T_max = p_int(tags = c("train", "required")),
+      eta_min = p_dbl(default = 0, tags = "train"),
+      last_epoch = p_int(default = -1, tags = "train"),
+      verbose = p_lgl(default = FALSE, tags = "train")
+    ),
+    id = "lr_cosine_annealing",
+    label = "Learning Rate Scheduler using Cosine Annealing",
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:77]:
```suggestion
    label = "Cosine Annealing LR Scheduler",
```

File: R/CallbackSetLRScheduler.R [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+mlr3torch_callbacks$add("lr_lambda", function() {
+  TorchCallback$new(
+    callback_generator = CallbackSetLRScheduler,
+    param_set = ps(
+      lr_lambda = p_uty(tags = c("train", "required"), custom_check = function(x) check_class_or_list(x, "function")),
+      last_epoch = p_int(default = -1, tags = "train"),
+      verbose = p_lgl(default = FALSE, tags = "train")
+    ),
+    id = "lr_scheduler",
+    label = "Learning Rate Scheduler using Multplication by a Function",
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:93]:
```suggestion
    label = "Multiplicative LR Scheduler",
```

File: R/TorchDescriptor.R [hunk 23]
```diff
@@ -37,7 +37,9 @@ TorchDescriptor = R6Class("TorchDescriptor",
     #' @template param_packages
     #' @template param_label
     #' @template param_man
-    initialize = function(generator, id = NULL, param_set = NULL, packages = NULL, label = NULL, man = NULL) {
+    #' @param additional_args (`any`)\cr
```
Comment (sebffischer) [inline R/TorchDescriptor.R:40]:
```suggestion
    #' @param additional_args (`list()`)\cr
```

File: R/TorchDescriptor.R [hunk 24]
```diff
@@ -63,6 +65,7 @@ TorchDescriptor = R6Class("TorchDescriptor",
       self$id = assert_string(id %??% class(generator)[[1L]], min.chars = 1L)
       self$label = assert_string(label %??% self$id, min.chars = 1L)
       self$packages = assert_names(unique(union(packages, c("torch", "mlr3torch"))), type = "strict")
+      private$.additional_args = additional_args
```
Comment (sebffischer) [inline R/TorchDescriptor.R:68]:
```suggestion
      private$.additional_args = assert_list(additional_args)
```

File: tests/testthat/test_CallbackSetLRScheduler.R [hunk 25]
```diff
@@ -0,0 +1,66 @@
+test_that("autotest", {
+  cb = t_clbk("lr_cosine_annealing", T_max = 10)
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetLRScheduler.R:2]: just run this test with all callbacks

**Review by sebffischer — CHANGES_REQUESTED** [review]

File: R/CallbackSetLRScheduler.R [hunk 26]
```diff
@@ -132,7 +123,7 @@ mlr3torch_callbacks$add("lr_one_cycle", function() {
       verbose = p_lgl(default = FALSE, tags = "train")
     ),
     id = "lr_one_cycle",
-    label = "Learning Rate Scheduler using 1cycle",
+    label = "1cyle Learning Rate Scheduler",
```
Comment (sebffischer) [inline R/CallbackSetLRScheduler.R:126]:
```suggestion
    label = "1cyle LR Scheduler",
```

**Review by sebffischer — APPROVED** [review]
